internal/fwschemadata/data_set_at_path.go

```diff
@@ -158,13 +158,27 @@ func (d Data) SetAtPathTransformFunc(ctx context.Context, path path.Path, tfVal
 	}
 
 	if parentValue.IsNull() || !parentValue.IsKnown() {
-		// TODO: This will break when DynamicPsuedoType is introduced.
-		// tftypes.Type should implement AttributePathStepper, but it currently does not.
-		// When it does, we should use: tftypes.WalkAttributePath(p.Raw.Type(), parentPath)
-		// Reference: https://github.com/hashicorp/terraform-plugin-go/issues/110
 		parentType := parentAttrType.TerraformType(ctx)
 		var childValue interface{}
 
+		// If the parent type from the schema is DynamicPseudoType, we need to use the type from the concrete value (if available)
+		if parentType.Is(tftypes.DynamicPseudoType) {
+			valueType := parentValue.Type()
+
+			// If the value is also DynamicPseudoType (i.e, null or unknown), we cannot perform an upsert
+			if valueType.Is(tftypes.DynamicPseudoType) {
+				diags.AddAttributeError(
+					parentPath,
+					d.Description.Title()+" Write Error",
+					"An unexpected error was encountered trying to write an attribute to the "+d.Description.String()+". This is always an error in the provider. Please report the following to the provider developer:\n\n"+
+						"Parent value is tftypes.DynamicPseudoType, which cannot be upserted to.",
+				)
+				return nil, diags
+			}
+
+			parentType = valueType
+		}
+
```

Comment:
Not directly relevant to this line, but further down in this function, on line 197, the `UpsertChildTerraformValue` function currently does not support tuples, which we "technically" have opened the door for...

It feels like I should implement it with this PR, any opinion on yay or nay?
https://github.com/hashicorp/terraform-plugin-framework/blob/a01f387340cea4d52f5cc3c890d540b19cb43f00/internal/fwschemadata/tftypes_value.go#L68-L69

Reply:
Upserting with dynamic values will no longer be possible due to the new attribute pathing, so no tuple support is needed unless we introduce a statically type `TupleAttribute` 🥳